Fix race condition with exec and resize

hqhq opened the pull request:
When I use `docker exec -ti test ls`, I got error:

```
ERRO[0035] Handler for POST /v1.23/exec/9677ecd7aa9de96f8e9e667519ff266ad26a5be80e80021a997fff6084ed6d75/resize returned error: bad file descriptor
```

It's because `POST /exec/<id>/start` and
`POST /exec/<id>/resize` are asynchronous, it is
possible that exec process finishes and ternimal
is closed before resize. Then `console.Fd()` will
get a large invalid number and we got the above
error.

Fix it by adding synchronization between exec and
resize.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

hqhq commented:
I found this error when I was working on https://github.com/docker/docker/pull/21018
It can be reproduced easily, but after fix in #21018 , It's a lot harder to be reproduced, but the race condition is still there.

thaJeztah commented:
@hqhq can you look into the test failures?

```
19:22:38 ---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
19:22:39 daemon/exec/exec.go:118: arg c.ID for printf verb %d of wrong type: string
19:22:39 exit status 1
```

hqhq committed:
Fix race condition with exec and resize

When I use `docker exec -ti test ls`, I got error:
```
ERRO[0035] Handler for POST /v1.23/exec/9677ecd7aa9de96f8e9e667519ff266ad26a5be80e80021a997fff6084ed6d75/resize returned error: bad file descriptor
```

It's because `POST /exec/<id>/start` and
`POST /exec/<id>/resize` are asynchronous, it is
possible that exec process finishes and ternimal
is closed before resize. Then `console.Fd()` will
get a large invalid number and we got the above
error.

Fix it by adding synchronization between exec and
resize.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq commented:
@thaJeztah Fixed, have no idea about failed test cases for now :(

thaJeztah commented:
ping @jhowardmsft panic on Windows? https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/2487/console

```
11:10:04 
11:10:04 ----------------------------------------------------------------------
11:10:04 PANIC: check_test.go:30: DockerSuite.TearDownTest
11:10:04 
11:10:04 ... Panic: runtime error: index out of range (PC=0x45E085)
11:10:04 
11:10:04 c:/go/src/runtime/asm_amd64.s:472
11:10:04   in call32
11:10:04 c:/go/src/runtime/panic.go:426
11:10:04   in gopanic
11:10:04 c:/go/src/runtime/panic.go:15
11:10:04   in panicindex
11:10:04 github.com/docker/docker/integration-cli/_test/_obj_test/docker_utils.go:471
11:10:04   in deleteAllImages
11:10:04 check_test.go:33
11:10:04   in DockerSuite.TearDownTest
11:10:04 c:/go/src/runtime/asm_amd64.s:472
11:10:04   in call32
11:10:04 c:/go/src/reflect/value.go:435
11:10:04   in Value.call
11:10:04 c:/go/src/reflect/value.go:303
11:10:04   in Value.Call
11:10:04 C:/gopath/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:721
11:10:04   in suiteRunner.runFixture.func1
11:10:04 C:/gopath/src/github.com/docker/docker/vendor/src/github.com/go-check/check/check.go:666
11:10:04   in suiteRunner.forkCall.func1
11:10:04 c:/go/src/runtime/asm_amd64.s:1998
11:10:04   in goexit
11:10:04 
11:10:04 ----------------------------------------------------------------------
11:10:04 PANIC: docker_cli_build_test.go:2278: DockerSuite.TestBuildExpose
11:10:04 
11:10:04 ... Panic: Fixture has panicked (see related PANIC)
11:10:04 
11:10:04 ----------------------------------------------------------------------
11:10:04 MISS: docker_cli_build_test.go:2295: DockerSuite.TestBuildExposeMorePorts
```

thaJeztah commented:
ping @cpuguy83 @calavera ptal `<3`

calavera commented:
LGTM

runcom commented:
LGTM
